### PR TITLE
BIG-4022 - Add a `simhash-find-all` binary

### PR DESCRIPTION
**Pull request by dlecocq**
`simhash-find-all` reads a stream of 64-bit integers (of any endianness) and write a stream of pairs of 64-bit integers (of the same endianness) that match the `--distance` criteria. It uses `--blocks` blocks to perform this calculation.

@tammybailey @b4hand @lindseyreno 

@bfeigin @benkirzhner @koos42 -- does this look like it will work for you guys?

### CODE REVIEW

File: src/bin/simhash-find-all.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+#include "simhash.h"
+
+void usage(int argc, char** argv)
+{
+    std::cout << "usage: " << argv[0]
+              << " --blocks BLOCKS"
+              << " --distance DISTANCE"
+              << " --input INPUT"
+              << " --output OUTPUT"
+              << std::endl
```
Comment (b4hand) [inline src/bin/simhash-find-all.cpp]: Martin would say: there's no need to flush after every line. You can just use a literal newline character in your strings to end the lines.

File: src/bin/simhash-find-all.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            reinterpret_cast<const char*>(&(it->first)), sizeof(Simhash::hash_t));
+        stream.write(
+            reinterpret_cast<const char*>(&(it->second)), sizeof(Simhash::hash_t));
+    }
+    stream.flush();
+}
+
+int main(int argc, char **argv) {
+
+    std::string input(""), output("");
```
Comment (b4hand) [inline src/bin/simhash-find-all.cpp]: The default constructor for `std::string` is an empty string. There's no need to pass a constructor argument. This just causes additional work which the compiler can't avoid.

File: src/bin/simhash-find-all.cpp (hunk 1 above)
Comment (dlecocq) [inline src/bin/simhash-find-all.cpp]: Sure, but this is a usage string and so it's not a hot loop. Isn't `std::endl` more idiomatic unless it's performance-critical?

File: src/bin/simhash-find-all.cpp (hunk 2 above)
Comment (dlecocq) [inline src/bin/simhash-find-all.cpp]: Ok. This isn't a hot loop, but I'm fine omitting it.

File: src/bin/simhash-find-all.cpp (hunk 1 above)
Comment (b4hand) [inline src/bin/simhash-find-all.cpp]: I think either is fine from an idiomatic perspective. In this particular case, I think `\n` will make it a little less cluttered and readable.

**Pull request merged**